src/Runner/Runner.php

```diff
@@ -109,20 +115,92 @@ public function __construct(
         $this->cacheManager = $cacheManager;
         $this->directory = $directory ?: new Directory('');
         $this->stopOnViolation = $stopOnViolation;
+        $this->parallel = $parallel;
     }
 
     public function fix(): array
     {
-        $changed = [];
+        $finderIterator = $this->finder instanceof \IteratorAggregate
+            ? $this->finder->getIterator()
+            : $this->finder;
 
-        $finder = $this->finder;
-        $finderIterator = $finder instanceof \IteratorAggregate ? $finder->getIterator() : $finder;
         $fileFilteredFileIterator = new FileFilterIterator(
             $finderIterator,
             $this->eventDispatcher,
             $this->cacheManager
         );
 
+        if ($this->parallel > 1) {
+            return $this->fixParallel($fileFilteredFileIterator);
+        }
+
+        return $this->fixProcedural($fileFilteredFileIterator);
+    }
+
+    private function fixParallel(FileFilterIterator $fileFilteredFileIterator): array
+    {
+        if (false === \function_exists('pcntl_fork')) {
+            return $this->fixProcedural($fileFilteredFileIterator);
+        }
+
+        $changed = [];
+        $childProcesses = [];
+
+        $files = iterator_to_array($fileFilteredFileIterator);
+        $numFiles = \count($files);
+        $numPerBatch = (int) ceil($numFiles / $this->parallel);
+
+        foreach (array_chunk($files, $numPerBatch) as $batch) {
+            $childOutFilename = tempnam(sys_get_temp_dir(), 'php-cs-fixer');
+            $pid = pcntl_fork();
+            if (-1 === $pid) {
+                exit('could not fork');
+            }
+            if ($pid) {
+                $childProcesses[$pid] = $childOutFilename;
+            } else {
+                $changedBatch = [];
+
+                /** @var \SplFileInfo $file */
+                foreach ($batch as $file) {
+                    $fixInfo = $this->fixFile($file, $this->linter->lintFile($file->getRealPath()));
+
+                    // we do not need Tokens to still caching just fixed file - so clear the cache
+                    Tokens::clearCache();
+
+                    if ($fixInfo) {
+                        $name = $this->directory->getRelativePathTo($file->__toString());
+                        $changedBatch[$name] = $fixInfo;
+                    }
+                }
+
+                $output = '<?php'.PHP_EOL;
+                $output .= 'return '.var_export($changedBatch, true).';'.PHP_EOL;
+
+                file_put_contents($childOutFilename, $output);
```

Comment:
does the communication from children to parent process needed to be via file?

Reply:
Replying here also for https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5822#discussion_r679524110
Unfortunately, children and the master process don't share anything so I had to find an external solution to retrieve children's results.
I've found out also that there's an issue in my implementation since the `ErrorsManager` collect errors in the child processes and those collected values are not visible in the master (`RunnerTest::testThatFixInvalidFileReportsToErrorManager` will fail)

Reply:
then we would probably also have issues with not storing the cache,
ref `$this->cacheManager->setFile($name, $new);`